src/main/java/org/opensearch/flowframework/template/ProcessNode.java

```diff
@@ -0,0 +1,184 @@
+    }
+
+    /**
+     * Create this node linked to its executing process.
+     *
+     * @param id A string identifying the workflow step
+     * @param workflowStep A java class implementing {@link WorkflowStep} to be executed when it's this node's turn.
+     * @param input Input required by the node
+     */
+    public ProcessNode(String id, WorkflowStep workflowStep, WorkflowData input) {
+        this.id = id;
+        this.workflowStep = workflowStep;
+        this.input = input;
+    }
+
+    /**
+     * Returns the node's id.
+     * @return the node id.
+     */
+    public String id() {
+        return id;
+    }
+
+    /**
+     * Returns the node's workflow implementation.
+     * @return the workflow step
+     */
+    public WorkflowStep workflowStep() {
+        return workflowStep;
+    }
+
+    /**
+     * Returns the input data for this node.
+     * @return the input data
+     */
+    public WorkflowData input() {
+        return input;
+    }
+
+    /**
+     * Returns a {@link CompletableFuture} if this process is executing.
+     * Relies on the node having been sorted and executed in an order such that all predecessor nodes have begun execution first (and thus populated this value).
+     *
+     * @return A future indicating the processing state of this node.
+     * Returns {@code null} if it has not begun executing, should not happen if a workflow is sorted and executed topologically.
+     */
+    public CompletableFuture<WorkflowData> getFuture() {
+        return future;
+    }
+
+    /**
+     * Returns the predecessors of this node in the workflow.
+     * The predecessor's {@link #getFuture()} must complete before execution begins on this node.
+     *
+     * @return a set of predecessor nodes, if any.  At least one node in the graph must have no predecessors and serve as a start node.
+     */
+    public Set<ProcessNode> getPredecessors() {
+        return predecessors;
+    }
+
+    /**
+     * Sets the predecessor node.  Called by {@link TemplateParser}.
+     *
+     * @param predecessors The predecessors of this node.
+     */
+    void setPredecessors(Set<ProcessNode> predecessors) {
+        this.predecessors = Set.copyOf(predecessors);
+    }
+
+    /**
+     * Execute this node in the sequence. Initializes the node's {@link CompletableFuture} and completes it when the process completes.
+     *
+     * @return this node's future. This is returned immediately, while process execution continues asynchronously.
+     */
+    public CompletableFuture<WorkflowData> execute() {
+        this.future = new CompletableFuture<>();
+        CompletableFuture.runAsync(() -> {
+            List<CompletableFuture<WorkflowData>> predFutures = predecessors.stream().map(p -> p.getFuture()).collect(Collectors.toList());
+            if (!predecessors.isEmpty()) {
+                CompletableFuture<Void> waitForPredecessors = CompletableFuture.allOf(predFutures.toArray(new CompletableFuture<?>[0]));
```

Comment:
Why we do need a new future here? We can just add a timeout for `predFutures` to get populated.

Reply:
I'm using this for `allOf()`.  

Another option is to stick them all in a list and call `join()` on them.

Not sure what other option you are proposing.

Reply:
I missed `allOf()` due to scrolling. Looks good and let's keep `allOf()` only here as it allows us to continue our flow in a non-blocking way.